### ISSUE TITLE
The output layer of the seq2seq model is not shared between training and evaluation? Neither is it shared with the word embedding.

### ISSUE THREAD

**Issue by shrshore** [OpenNMT/OpenNMT-tf#97]
In the https://github.com/OpenNMT/OpenNMT-tf/blob/master/opennmt/models/sequence_to_sequence.py. The line 156 calls the `decode` method to decode the next step logits, without passing the parameter `output_layer`. While in the L181 and L194, the dynamic_decode and dynamic_decode_and search methods are called without `output_layer' parameters. So the output layer is not shared between training and evaluation/testing. Is it a bug or intended? Could someone give some explanation?

Another problem is that in the `build_output_layer', it calls `tf.layers.Dense' to project the input into logits, which cannot share transform parameters with the encoding word embeddings.

**Comment by guillaumekln** [comment]
For the first question, the output layer is correctly reusing the training weights. It is created under the same variable scope and the [`Estimator`](https://www.tensorflow.org/programmers_guide/estimators) framework we use for managing the model makes sure to set `reuse=True` when building the evaluation and inference graph.

For the second question, it is not required to share the embeddings and the projection layer. For example, we don't do that by default in the other OpenNMT implementations. However, we would like to support this as an option in the future.

**Comment by atebbifakhr** [comment]
I'm also dealing with this issue, I want to use `decoder.decode` and `decoder.dynamic_decode_and_search` together during training, but the second one creates another `tf.layers.Dense` and `tf.contrib.layers.layer_norm`.
